src/ralph/schemas/edx/browser.py

```diff
@@ -227,18 +219,94 @@ def validate_event_textbook_pdf(self, data, **kwargs):
+            return
+
+        event = self.parse_json_check_keys(
+            data, {"chapter", "page", "name", "thumbnail_title"}
+        )
+        self.check_chapter_page_name(data, event)
+        if not isinstance(event["thumbnail_title"], str):
+            raise ValidationError("thumbnail_title should be a string")
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_buttons_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `direction`: string (one of [`in`, `out`] or [`up`, `down`])
+        """
+        if data["name"] not in [
+            "textbook.pdf.zoom.buttons.changed",
+            "textbook.pdf.page.scrolled",
+        ]:
+            return
+
+        event = self.parse_json_check_keys(
+            data, {"chapter", "page", "name", "direction"}
+        )
+        self.check_chapter_page_name(data, event)
+        valid_directions = (
+            ["in", "out"] if data["name"][-7:] == "changed" else ["up", "down"]
+        )
+        if event["direction"] not in valid_directions:
+            raise ValidationError(
+                f"event.direction should be one of {valid_directions}"
+            )
+
+    @validates_schema
+    def validate_event_textbook_pdf_zoom_menu_changed(self, data, **kwargs):
+        """Check that the event is a parsable JSON object with 4 keys:
+        `name`: string, `page`: integer > 0, `chapter`: url,
+        `amaunt`: string
+        """
+        if data["name"] != "textbook.pdf.zoom.menu.changed":
+            return
+
+        event = self.parse_json_check_keys(data, {"chapter", "page", "name", "amaunt"})
+        self.check_chapter_page_name(data, event)
+        if event["amaunt"] not in BROWSER_EVENT_VALID_AMAUNT:
+            raise ValidationError(
+                f"amaunt should be one of {BROWSER_EVENT_VALID_AMAUNT}"
+            )
+
+    @validates("session")
+    def validate_session(self, value):  # pylint: no-self-use
+        """"Check session field empty or MD5_HASH_LEN chars long"""
+        if value != "" and len(value) != MD5_HASH_LEN:
+            raise ValidationError(
+                f"Session should be empty or {MD5_HASH_LEN} chars long (MD5)"
+            )
+
+    @staticmethod
+    def parse_json_check_keys(data, required_keys):
+        """Parse JSON string and check that it contains the keys in the
+        `keys_array` and no other keys
+        """
         try:
             event = json.loads(data["event"])
         except (json.JSONDecodeError, TypeError):
-            raise ValidationError("Event should contain a parsable json string")
-
-        required_keys = {"name", "page", "chapter"}
+            raise ValidationError("Event should contain a parsable JSON string")
         keys = set(event.keys())
-        if len(required_keys.intersection(keys)) != 3:
+        if keys != required_keys:
```

Comment:
I think this is not what we want, but rather:

```suggestion
        if len(required_keys.intersection(keys)) != len(required_keys):
```

WDYT?

Reply:
My concern here was that we might silently accept events having the required keys but also other keys that will be ignored in this case.
If this is not a problem, it's good for me, i will change that.

Reply:
From your experience, does this occurs?

Reply:
If I understood it right, the event field is created by the client then send to the django back-end and in the back-end there is no validation (it just tries to parse the json if it's a string)
I suggest the event field can't be trusted as it might contain any values.

Reply:
Ok, so feel free to be more stringent then.